todo-mvc/src/widgets/createTodoItem.ts

```diff
@@ -17,82 +17,125 @@ interface TodoItemState extends WidgetState, StatefulChildrenState {
 
-	checkbox.setState({
-		checked: todoItem.state.completed
+	checkbox.widget.setState({
+		checked: this.state.completed
 	});
 }
 
 const createTodoItem: TodoItemFactory = createWidget
-	.mixin(createParentMixin)
 	.mixin(createRenderableChildrenMixin)
-	.mixin(createStatefulChildrenMixin)
 	.mixin({
-		mixin: createParentMixin,
-		initialize(instance) {
-			const checkbox = createCheckboxInput({
-				state: {
-					id: 'checkbox',
-					classes: ['toggle']
-				},
-				listeners: {
-					change: () => { todoToggleComplete.do(instance.state); }
-				}
-			});
-
-			const button = createButton({
-				state: {
-					id: 'button',
-					classes: ['destroy']
-				},
-				listeners: {
-					click: () => { todoRemove.do(instance.state); }
-				}
-			});
-
-			const label = createWidget({
-				state: {
-					id: 'label'
-				},
-				listeners: {
-					dblclick: () => { todoEdit.do(instance.state); }
-				},
-				tagName: 'label'
-			});
-
-			const editInput = createFocusableTextInput({
-				state: {
-					id: 'editInput',
-					classes: ['edit']
-				},
-				listeners: {
-					blur: (event) => { todoSave.do({state: instance.state, event}); },
-					keyup: (event) => { todoEditInput.do({state: instance.state, event}); }
-				}
-			});
-
-			instance.append([checkbox, button, label, editInput]);
-			instance.on('statechange', manageChildren);
+		mixin: createStatefulChildrenMixin,
+		initialize(instance, options) {
+			instance
+				.createChildren({
+					checkbox: {
+						factory: createCheckboxInput,
+						options: {
+							state: {
+								classes: [ 'toggle' ]
+							},
+							listeners: {
+								change: () => { todoToggleComplete.do(instance.state); }
+							}
+						}
+					},
+					button: {
+						factory: createButton,
+						options: {
+							state: {
+								classes: [ 'destroy' ]
+							},
+							listeners: {
+								click: () => { todoRemove.do(instance.state); }
+							}
+						}
+					},
+					label: {
+						factory: createWidget,
+						options: {
+							listeners: {
+								dblclick: () => { todoEdit.do(instance.state); }
+							},
+							tagName: 'label'
+						}
+					},
+					editInput: {
+						factory: createFocusableTextInput,
+						options: {
+							state: {
+								classes: [ 'edit' ]
+							},
+							listeners: {
+								blur: (event: Event) => { todoSave.do({state: instance.state, event}); },
+								keyup: (event: Event) => { todoEditInput.do({state: instance.state, event}); }
+							}
+						}
+					}
+				})
+				.then((children: TodoItemChildren) => {
+					/* TODO: We are only using the label.widget but we are storing label: { id, widget }, is
+					 * that necessary? */
+					childrenMap.set(instance, children);
+					instance.on('statechange', manageChildren);
+
+					/* We have missed the widgets initial state change because we created the widgets async
+					 * so we should emit a statechange event just to ensure we get the logic we expect
+					 * to be handled */
+					instance.emit({
```

Comment:
@kitsonk no this doesn't block this PR at all 👍. more of a discussion really. I'll reply to one of the questions above inline here as it doesn't exist in context in the issue you've just opened:
> . Also, if we indiscriminately store every event just in case a listeners comes along later, we would run into a memory issue.

In this example (and this is actually true for a lot of event types), you would use a sliding buffer with a size of 1, you only care about the latest event for state.